### ISSUE TITLE
IO.read(bytes) behaves differently to MRI on Windows

### ISSUE THREAD

**Issue by deployable** [jruby/jruby#2308]
From issue: https://github.com/djberg96/ptools/pull/23#issuecomment-66496525 @djberg96 found a behaviour difference in JRuby to MRI

MRI [documents IO#read behaviour](http://www.ruby-doc.org/core-1.9.3/IO.html#method-i-read)
    > If length is a positive integer, it try to read length bytes without any conversion (binary mode).

It appears JRuby doesn't track  MRI behaviour on Windows when specifying a byte length to read  (`f.read(bytes)`). MRI changes to binary mode. JRuby still does the line end conversion. 

`ruby 1.9.3p551 (2014-11-13) [i386-mingw32]`

```
C:\>c:\lang\ruby-1.9.3-p551-i386-mingw32\bin\pry
[1] pry(main)> File.open('test_slashrn','rb'){|f| f.read }
=> "something\r\notherthing\r\nthirdthing\r\n"
[2] pry(main)> File.open('test_slashrn'){|f| f.read }
=> "something\notherthing\nthirdthing\n"
[3] pry(main)> File.open('test_slashrn'){|f| f.read(35) }
=> "something\r\notherthing\r\nthirdthing\r\n"
```

`jruby 1.7.17 (1.9.3p392) 2014-12-09 fafd1a7 on Java HotSpot(TM) 64-Bit Server VM 1.7.0_07-b10 +jit [Windows 7-amd64]`

```
C:\>c:\lang\jruby-1.7.17\bin\pry
←[0G[1] pry(main)> File.open('test_slashrn','rb'){|f| f.read }
=> "something\r\notherthing\r\nthirdthing\r\n"
[2] pry(main)> File.open('test_slashrn'){|f| f.read }
=> "something\notherthing\nthirdthing\n"
[3] pry(main)> File.open('test_slashrn'){|f| f.read(35) }
=> "something\notherthing\nthirdthing\n"
```

This isn't an issue on Linux as neither MRI nor JRuby attempt conversion from `\r\n` to `\n` on the `read`.

**Comment by headius** [comment]
Master does work on Windows now, so if you can test 9.0.0.0.pre2 and report back that would be fantastic.

The problem here on both 1.7 and 9k is that we can't just rely on system-level O_BINARY and O_TEXT; there's no way to set those in Java NIO channels. MRI's logic does not manually do any crlf conversion; it merely sets the appropriate flag, and Windows POSIX APIs do the translation for it.

In 9k, I think we'll need to figure out a way to set binmode/textmode, or introduce a channel wrapper that can do the same. In 1.7, we need to modify the logic to tell our crlf-converting wrapper not to convert for binary reads.

This is not going to get fixed for 1.7.20, so I'm bumping it to 21.

**Comment by enebo** [comment]
This is broken on 1.7 and 9k in that we do not return proper crlf or lack thereof.  On 1.7, we only fail in the third test (read(amount)) and we return as \n vs \r\n.  On 9k, we also fail the third test similarly plus we fail the first test and return \n.  I am punting because I do not want to dig into guts of IO in 1.7 this close to release.

**Comment by headius** [comment]
Not going to be fixed in 1.7. This works properly on Windows with JRuby 9k. Please upgrade.
